newapi/ooniapi/measurements.py

```diff
@@ -1198,6 +1708,175 @@ def coalsum(name):
         return jsonify({"v": 0, "error": str(e)})
 
 
+def _clickhouse_aggregation(
+    resp_format,
+    since,
+    until,
+    inp,
+    domain,
+    category_code,
+    probe_cc,
+    probe_asn,
+    test_name,
+    axis_x,
+    axis_y,
+):
+    log = current_app.logger
+    dimension_cnt = int(bool(axis_x)) + int(bool(axis_y))
+    cacheable = until and until < datetime.now() - timedelta(hours=72)
+    cacheable = False  # FIXME
+
+    # Assemble query
+    colnames = [
+        "anomaly_count",
+        "confirmed_count",
+        "failure_count",
+        "measurement_count",
+    ]
+    cols = [
+        sql.text("countIf(anomaly = 't') AS anomaly_count"),
+        sql.text("countIf(confirmed = 't') AS confirmed_count"),
+        sql.text("countIf(msm_failure = 't') AS failure_count"),
+        sql.text("COUNT(*) AS measurement_count"),
+    ]
+    table = sql.table("fastpath")
+    where = []
+    query_params = {}
+
+    if domain:
+        where.append(sql.text("domain = :domain"))
+        query_params["domain"] = domain
+
+    if inp:
+        where.append(sql.text("input = :input"))
+        query_params["input"] = inp
+
+    if category_code:
+        where.append(sql.text("category_code = :category_code"))
+        query_params["category_code"] = category_code
+        log.error("@"*66)
+
+    if probe_cc:
+        where.append(sql.text("probe_cc = :probe_cc"))
+        query_params["probe_cc"] = probe_cc
+
+    if probe_asn is not None:
+        if probe_asn.startswith("AS"):
+            probe_asn = probe_asn[2:]
+        probe_asn = int(probe_asn)
+        where.append(sql.text("probe_asn = :probe_asn"))
+        query_params["probe_asn"] = probe_asn
+
+    if since:
+        where.append(sql.text("measurement_start_time > :since"))
+        query_params["since"] = since
+
+    if until:
+        where.append(sql.text("measurement_start_time <= :until"))
+        query_params["until"] = until
+
+    if test_name:
+        assert test_name in TEST_NAMES
+        where.append(sql.text("test_name = :test_name"))
+        query_params["test_name"] = test_name
+
+    if axis_x and axis_x == axis_y:
+        assert 0
+
+    if axis_x and axis_x == axis_y:
+        raise ValueError()
+
+    def group_by_date(since, until, cols, colnames, group_by):
+        if since and until:
+            delta = abs((until - since).days)
+        else:
+            delta = 0
+
+        # lim = 100
+        # if delta > lim * 30:
+        #     granularity = ("year", "toYear")
+        # elif delta > lim * 7:
+        #     granularity = ("month", "toMonth")
+        # elif delta > lim:
+        #     granularity = ("week", "toWeek")
+        if delta < 2:
```

Comment:
I am not sure it's a good idea for the backend to be doing this type of dynamic/magic scaling of the granularity of the returned data. I would rather it's the user explicitly asking for a higher or lower level of granularity and them passing it to the backend via the UI.

Reply:
This is why is mostly commented out. Users might be unaware of what constitutes the right granularity and poor choices impact the API; also we might want to be able to tune the granularity centrally over time to optimize performance.

However we could let the API pick the preferred granularity by default, to make the MAT user friendly, and also allow users to require a fixed granularity. In the second case the API could simply refuse overly heavy queries.

Reply:
Dynamic scaling removed.